packages/sdk/src/cross-chain-messenger.ts

```diff
@@ -1267,12 +1260,168 @@ export class CrossChainMessenger {
+        getContractInterfaceBedrock('FaultDisputeGame'),
+        this.l1SignerOrProvider
+      )
+
+      // Check the game status.
+      const status = await game.status()
+      if (status === 1) {
+        // If status is CHALLENGER_WINS then it's no good.
+        continue
+      } else if (status === 2) {
+        // If status is DEFENDER_WINS then it's a valid proof.
+        return {
+          ...provenWithdrawal,
+          proofSubmitter,
+        }
+      } else if (status > 2) {
+        // Shouldn't happen in practice.
+        throw new Error('got invalid game status')
+      }
+
+      // Otherwise we're IN_PROGRESS.
+      // Grab the block number from the extra data. Since this is not a standardized field we need
+      // to be defensive and assume that the extra data could be anything. If the extra data does
+      // not decode properly then we just skip this game.
+      const extraData = await game.extraData()
+      let l2BlockNumber: number
+      try {
+        ;[l2BlockNumber] = ethers.utils.defaultAbiCoder.decode(
+          ['uint256'],
+          extraData
+        )
+      } catch (err) {
+        // Didn't decode properly, bad game.
+        continue
+      }
+
+      // Finally we check if the output root is valid. If it is, then we can return the proven
+      // withdrawal. If it isn't, then we act as if this proof does not exist because it isn't
+      // useful for finalizing the withdrawal.
+      if (await this.isValidOutputRoot(await game.rootClaim(), l2BlockNumber)) {
+        return {
+          ...provenWithdrawal,
+          proofSubmitter,
+        }
+      }
+    }
+
+    // Return null if we didn't find a valid proof.
+    return null
+  }
+
+  /**
+   * Checks whether a given root claim is valid. Uses the L2 node that the SDK is connected to
+   * when verifying the claim. Assumes that the connected L2 node is honest.
+   *
+   * @param outputRoot Output root to verify.
+   * @param l2BlockNumber L2 block number the root is for.
+   * @returns Whether or not the root is valid.
+   */
+  public async isValidOutputRoot(
+    outputRoot: string,
+    l2BlockNumber: number
+  ): Promise<boolean> {
+    // Use the cache if we can.
+    const cached = this._outputCache.find((other) => {
+      return other.root === outputRoot
+    })
+
+    // Skip if we can use the cached.
+    if (cached) {
+      return cached.valid
+    }
+
+    // If the cache ever gets to 10k elements, clear out the first half. Works well enough
+    // since the cache will generally tend to be used in a FIFO manner.
+    if (this._outputCache.length > 10000) {
+      this._outputCache = this._outputCache.slice(5000)
+    }
+
+    // We didn't hit the cache so we're going to have to do the work.
+    try {
+      // Make sure this is a JSON RPC provider.
+      const provider = toJsonRpcProvider(this.l2Provider)
+
+      // Grab the block and storage proof at the same time.
+      const [block, proof] = await Promise.all([
```

Comment:
Nit: Don't bother changing this IMO it's fine as is but in general I've been pretty exclusively only using [`Promise.allSettled`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled) rather than Promise.all because it allows you to check both promises for an error rather than them racing to which promise errors first. Just calling it out so anyone reading this considers using Promise.allSettled in future: